lib/middleware/operation.js

```diff
@@ -32,55 +36,95 @@ function findPropertyOperations ({ types, method, term, resource }) {
   return operations
 }
 
-function factory (api) {
-  return async (req, res, next) => {
-    if (!req.hydra.resource) {
-      return next()
+function mapOperations ({ api, res, method }) {
+  return res.locals.hydra.resources.flatMap((resource) => {
+    let moreOperations
+    if ('property' in resource) {
+      moreOperations = findPropertyOperations({ resource, method, api }).toArray()
+    } else {
+      const types = clownface({ ...api, term: [...resource.types] })
+      moreOperations = findClassOperations(types, method).toArray()
+    }
+
+    if (moreOperations.length) {
+      return moreOperations.map(operation => ({ resource, operation }))
     }
 
+    return [{ resource, operation: null }]
+  })
+}
+
+function findOperation (operationMiddleware) {
+  return async function (req, res, next) {
     const method = req.method === 'HEAD' ? 'GET' : req.method
-    const types = clownface({ ...api, term: [...req.hydra.resource.types] })
 
-    let operations
-    if (req.hydra.term.equals(req.hydra.resource.term)) {
-      // only look for direct operation when the root resource is requested
-      operations = findClassOperations(types, method)
+    let operations = mapOperations({ api: req.hydra.api, res, method })
+      .filter(({ operation }) => operation)
+
+    if (operationMiddleware) {
+      operations = operationMiddleware(operations, { req, res })
+    }
+
+    if (!operations.length) {
+      return next()
     }
-    if (!operations) {
-      // otherwise try finding the operation by property usage
-      operations = findPropertyOperations({ types, method, term: req.hydra.term, resource: req.hydra.resource })
+
+    if (operations.length > 1) {
+      error('Multiple operations found')
+      return next(new Error('Ambiguous operation'))
```

Comment:
400 range is intended for client errors. Failure to find the right operations is 99% going to be a server issue.

Reply:
I think most of the time it would be a missing or unclear rdf:type in a request body. But maybe it's better that the custom middleware explicitly sets that status code for such cases.